src/imputors/drop.jl

```diff
@@ -36,38 +46,134 @@ NOTES (or premature optimizations):
+
+Finds the missing rows in the table and deletes them.
+
+# Arguments
+* `imp::DropObs`: this `Imputor` method
+* `table`: a type that implements the Tables API.
+
+# Returns
+* our data with the missing rows removed.
+"""
+function impute!(imp::DropObs, table)
+    imp.context() do c
+        @assert istable(table)
+        rows = Tables.rows(table)
+
+        # Unfortunately, we'll need to construct a new table
+        # since Tables.rows is just an iterator
+        table = Iterators.filter(rows) do r
+            !any(x -> ismissing(c, x), propertyvalues(r))
+        end |> materializer(table)
+
+        return table
+    end
 end
 
+
 """
-    impute!(imp::Drop, ctx::Context, data::DataFrame)
+    DropVars <: Imputor
+
 
-Finds the missing rows in the `DataFrame` and deletes them.
+Removes missing values from the `AbstractArray` or `Tables.table` provided.
 
-NOTE: this isn't quite as fast as `dropnull` in `DataFrames`s as we're using an arbitrary
-`missing` function rather than using the precomputed `dt.isnull` vector of bools.
+# Fields
+* `context::AbstractContext`: A context which keeps track of missing data
+  summary information
+"""
+struct DropVars <: Imputor
+    context::AbstractContext
+end
+
+"""DropVars(; context=Context()) -> DropVars"""
+DropVars(; context=Context()) = DropVars(context)
+
+"""
+    impute!(imp::DropVars, data::AbstractMatrix)
+
+Finds columns in the matrix with too many missing values and uses a mask (Vector{Bool}) to
+return the `data` with those columns removed. Unfortunately, the mask approach
+requires copying the matrix.
 
 # Arguments
-* `imp::Drop`: this `Imputor` method
-* `ctx::Context`: contextual information for missing data
-* `data::DataFrame`: the data to impute
+* `imp::DropVars`: this `Imputor` method
+* `data::AbstractMatrix`: the data to impute
 
 # Returns
-* `DataFrame`: our data with the missing rows removed.
+* `AbstractMatrix`: a new matrix with missing columns removed
 """
-function impute!(imp::Drop, ctx::Context, data::DataFrame)
-    ctx.num = size(data, 1)
-    m = typeof(data).name.module
-    m.deleterows!(data, findall(r -> ismissing(ctx, r), m.eachrow(data)))
+function impute!(imp::DropVars, data::AbstractMatrix)
+    mask = map(1:size(data, 2)) do i
+        try
+            imp.context() do c
+                for j in 1:size(data, 1)
+                    ismissing(c, data[j, i])
+                end
+            end
+            return true
+        catch e
+            if isa(e, ImputeError)
+                return false
+            else
+                rethrow(e)
+            end
+        end
+    end
+
+    data = data[:, mask]
     return data
 end
+
+"""
+    impute!(imp::DropVars, table)
+
+Find remove columns in the table with too many missing elements.
+
+# Arguments
+* `imp::DropVars`: this `Imputor` method
+* `table`: a type that implements the Tables API.
+
+# Returns
+* our data with the missing columns removed.
+"""
+function impute!(imp::DropVars, table)
```

Comment:
I kind of want to define:
`DropDim{N}` and then just do 
`const DropVars=DropOnDim{1}`
`const DropObs=DropOnDim{2}`

Which would also mean that during transition one has the option of basically specifying which dimension is observations, for a softer upgrade path.
As the mansing of `DropVars` and `DropObs` might flip, but `DropOnDim` would stay the same.

Reply:
The issues is that it'll still be a breaking change for all the other Imputors. We could have the `impute` calls take a `dim` kwarg, but that wouldn't work nicely with the generated imputation methods that pass the kwargs to the Imputor constructor. Similarly, this convention only applies to imputing for a matrix.

Reply:
Introduced a `vardim` field in each `Imputor` that'll be ignored when the data isn't a matrix. If folks don't like the breaking behaviour then they can call `Impute.dropobs(; vardim=2)` or `Impute.dropvars(; vardim=2)`.